### PR TITLE
typesafe recursive metadata

### PR DESCRIPTION
**Pull request by airhorns**
This PR adapts @airhorns work to use model metadata fetched from the api as a array of metadata and then creating the circular references for `FieldMetadata[]` types; most of the work here is getting the mock metadata to look right in the tests; but I think a good 🎩 is still warrented cc @MillanWangGadget

### CODE REVIEW

File: packages/react/src/use-table/helpers.tsx [hunk 1]
```diff
@@ -336,27 +336,32 @@ const maybeGetDefaultDisplayFieldMetadata = (fieldMetadataTree: TableSpec["field
   }
 };
 
-export const fieldMetadataArrayToFieldMetadataTree = (fieldMetadataArray: FieldMetadataFragment[]) => {
+/**
+ * This turns the flat list of field metadata returned by the backend into a nested tree of field metadata.
+ * The paths in the object are precomputed to a fixed max depth, for easy lookups later in the table code
+ */
+export const fieldMetadataArrayToFieldMetadataTree = (fieldMetadataArray: FieldMetadata[], depth = 0) => {
```
Comment (infiton) [inline packages/react/src/use-table/helpers.tsx:343]: I didn't want to mess with the useTable stuff too much, this method builds out a tree that has the possible select paths that a table can take, we have to stop pre-expanding these paths out at a certain depth, I think 3 is sufficient for how AutoTable works

**Review by MillanWangGadget — COMMENTED** [review]
- 🟡 This change breaks all `AutoForm`s that have relationship fields due to `defaultDisplayField` no longer being included in the related model for relationship field configurations
  - `packages/react/src/auto/hooks/useRelatedModelOptions.tsx` has `defaultDisplayField.apiIdentifier` but defaultDisplayField is undefined
    - Easiest solution would be to bring back `defaultDispalyField`
    - Otherwise, this area would need to take the namespacedApiId of the related model and do a lookup in `modelAndRelatedModels`

**Comment by infiton** [comment]
> his change breaks all AutoForms that have relationship fields due to defaultDisplayField no longer being

good thing cypress caught all that; I think I've fixed it now

**Comment by airhorns** [comment]
Thanks for powering through on the tests and mocks Kyle 💖 ! And also just to confirm, from an API changes standpoint this is safe to ship because the new meta endpoints are already live in prod and not gated behind flags or anything, they were just net new 👍

**Review by MillanWangGadget — APPROVED** [review]
- 🍏 Confirmed that a circular relationship setup like this properly renders autoForms and that the metadata properly fetches unique models in `modelAndRelatedModels`
  - a belongs to b
  - b belongs to c
  - c belongs to a
- 🍏 Confirmed that self reference relationships work in autoform as well with these metadata updates

File: packages/react/src/metadata.tsx [hunk 2]
```diff
@@ -99,16 +78,9 @@ const FieldMetadataFragment = graphql(/* GraphQL */ `
         isJoinModelHasManyField
         relatedModel {
           key
+          name
```
[the remark this answers is not in the document]
Comment (infiton) [reply packages/react/src/metadata.tsx:81]: nope I think that's an oversight; I'll add it to both

**Pull request merged**